engines/docker/video.go

```diff
@@ -0,0 +1,51 @@
+package dockerengine
+
+import (
+	"path/filepath"
+	"regexp"
+
+	"github.com/pkg/errors"
+	funk "github.com/thoas/go-funk"
+)
+
+type device struct {
+	path    string
+	claimed bool
+}
+
+type videoDeviceManager struct {
+	devices []device
+}
+
+func newVideoDeviceManager() (*videoDeviceManager, error) {
```

Comment:
I would suggest something like:

```go
package dockerengine

import (
	"errors"
	"sync"

	"github.com/taskcluster/taskcluster-worker/runtime/atomics"
)

type deviceHandle interface {
	Path() string
	Release()
}

type videoDeviceManager struct {
	m       sync.Mutex
	devices []string
}

// newVideoDeviceManager creates a new videoDeviceManager
func newVideoDeviceManager() (*videoDeviceManager, error) {
	// TODO: Find or create devices
	devices := []string{"/dev/video0"}
	return &videoDeviceManager{devices: devices}, nil
}

// GetVideoDevice returns an idle video device
func (vdm *videoDeviceManager) GetVideoDevice() deviceHandle {
	vdm.m.Lock()
	defer vdm.m.Unlock()

	// ensure we have enough devices
	if len(vdm.devices) < 1 {
		panic(errors.New("insufficent number of video devices available"))
	}

	// Create a device handle, ensuring we can't release twice
	vdh := &videoDeviceHandle{owner: vdm}
	vdh.path, vdm.devices = vdm.devices[0], vdm.devices[1:]

	return vdh
}

// Dispose releases all resources assocated with the videoDeviceManager
func (vdm *videoDeviceManager) Dispose() error {
	vdm.m.Lock()
	defer vdm.m.Unlock()

	return nil // TODO: remove devices if any was created
}

type videoDeviceHandle struct {
	owner    *videoDeviceManager
	path     string
	released atomics.Once
}

func (vdh *videoDeviceHandle) Path() string {
	if vdh.released.IsDone() {
		panic(errors.New("videoDeviceHandle.Path() was called after release"))
	}
	return vdh.path
}

func (vdh *videoDeviceHandle) Release() string {
	ok := vdh.released.Do(func() {
		vdh.owner.m.Lock()
		defer vdh.owner.m.Unlock()

		vdh.owner.devices = append(vdh.owner.devices, vdh.path)
	})
	if !ok {
		debug("WARNING: harmless double release() of videoDeviceHandle")
	}
}

```

In the sandbox and resultset code you just release() the `deviceHandle` the same way we release `imageHandle` and pass it around the same way.

While techincally all this is in the same package, I only use methods starting with capital outside this file... the only exception being: `newVideoDeviceManager()`. This makes it easier to track life-cycles, and it is a motivation for moving device management into separate package, though that's a lot less important..